File: pine-go/internal/dataframe/column_write_test.go
package dataframe

import (
	"fmt"
	"strings"
	"testing"

	"github.com/Liam0205/pineapple/pine-go/internal/types"
)

// Batch column write (SetItemColumnFloat64) semantics: adopt on column
// store, scatter on row store, length check, NaN validation parity,
// ordering vs per-element writes.

func TestSetItemColumnFloat64BothModes(t *testing.T) {
	for _, tm := range testModes {
		t.Run(tm.name, func(t *testing.T) {
			items := []map[string]any{{"id": "a"}, {"id": "b"}, {"id": "c"}}
			f := NewFrame(tm.mode, nil, items)
			out := types.NewOperatorOutput()
			out.SetItemColumnFloat64("score", []float64{1.5, 2.5, 3.5})
			if err := f.ApplyOutput(out, "op", false); err != nil {
				t.Fatal(err)
			}
			for i, want := range []float64{1.5, 2.5, 3.5} {
				if got := f.Item(i, "score"); got != want {
					t.Errorf("Item(%d)=%v want %v", i, got, want)
				}
			}
			// Result projection includes the new field on every row.
			r := f.ToResult(nil, []string{"id", "score"})
			for i := range r.Items {
				if _, ok := r.Items[i]["score"]; !ok {
					t.Errorf("item %d missing score in result", i)
				}
			}
		})
	}
}

func TestSetItemColumnFloat64LengthMismatch(t *testing.T) {
	for _, tm := range testModes {
		t.Run(tm.name, func(t *testing.T) {
			f := NewFrame(tm.mode, nil, []map[string]any{{"id": "a"}, {"id": "b"}})
			out := types.NewOperatorOutput()
			out.SetItemColumnFloat64("score", []float64{1.0}) // wrong length
			err := f.ApplyOutput(out, "op", false)
			if err == nil {
				t.Fatal("expected length-mismatch error")
			}
			if !strings.Contains(err.Error(), "does not match item count 2") {
				t.Errorf("unexpected error: %v", err)
			}
		})
	}
}

func TestSetItemColumnFloat64NaNValidation(t *testing.T) {
	nan := func() float64 { var z float64; return 0 / z }()
	for _, tm := range testModes {
		t.Run(tm.name, func(t *testing.T) {
			f := NewFrame(tm.mode, nil, []map[string]any{{"id": "a"}, {"id": "b"}})
			out := types.NewOperatorOutput()
			out.SetItemColumnFloat64("score", []float64{1.0, nan})
			err := f.ApplyOutput(out, "op", false)
			if err == nil {
				t.Fatal("expected NaN validation error")
			}
			// Same first-error message shape as the per-element path.
			want := `item[1] write: field "score": NaN/Inf is not a valid JSON value`
			if err.Error() != want {
				t.Errorf("error = %q, want %q", err.Error(), want)
			}
		})
	}
}

func TestSetItemColumnFloat64OverridesPerElement(t *testing.T) {
	for _, tm := range testModes {
		t.Run(tm.name, func(t *testing.T) {
			f := NewFrame(tm.mode, nil, []map[string]any{{"id": "a"}, {"id": "b"}})
			out := types.NewOperatorOutput()
			out.SetItem(0, "score", 99.0) // per-element first
			out.SetItemColumnFloat64("score", []float64{1.0, 2.0})
			if err := f.ApplyOutput(out, "op", false); err != nil {
				t.Fatal(err)
			}
			// Column write applies after per-element → wins on collision.
			if got := f.Item(0, "score"); got != 1.0 {
				t.Errorf("Item(0)=%v want 1.0 (column write wins)", got)
			}
		})
	}
}

func TestSetItemColumnFloat64AdoptZeroCopy(t *testing.T) {
	// Column store must adopt the slice (typed fast-path readable,
	// zero-copy) — pinned via the read view aliasing the written slice.
	items := []map[string]any{{"id": "a"}, {"id": "b"}}
	f := newColumnFrame(nil, items)
	vals := []float64{1.0, 2.0}
	out := types.NewOperatorOutput()
	out.SetItemColumnFloat64("score", vals)
	if err := f.ApplyOutput(out, "op", false); err != nil {
		t.Fatal(err)
	}
	view, ok := f.ItemColumnFloat64View("score", 0, 2)
	if !ok {
		t.Fatal("adopted column must serve the typed read view")
	}
	if &view[0] != &vals[0] {
		t.Error("expected zero-copy adoption (view aliases the written slice)")
	}
}

func TestSetItemColumnFloat64RowColumnParity(t *testing.T) {
	mk := func(mode StorageMode) Frame {
		return NewFrame(mode, nil, []map[string]any{
			{"id": "a", "old": 1.0}, {"id": "b", "old": 2.0},
		})
	}
	apply := func(f Frame) {
		out := types.NewOperatorOutput()
		out.SetItemColumnFloat64("norm", []float64{0.25, 0.75})
		if err := f.ApplyOutput(out, "op", false); err != nil {
			t.Fatal(err)
		}
	}
	row, col := mk(StorageModeRow), mk(StorageModeColumn)
	apply(row)
	apply(col)
	fields := []string{"id", "old", "norm"}
	r1, r2 := row.ToResult(nil, fields), col.ToResult(nil, fields)
	for i := range r1.Items {
		if fmt.Sprint(r1.Items[i]) != fmt.Sprint(r2.Items[i]) {
			t.Errorf("item %d mismatch: %v vs %v", i, r1.Items[i], r2.Items[i])
		}
	}
}

File: pine-java/src/test/java/page/liam/pine/ColumnWriteTest.java
package page.liam.pine;

import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Batch column write (setItemColumnDouble) semantics: adopt on column
 * store, scatter on row store, length check, NaN validation parity,
 * ordering vs per-element writes. Mirrors pine-go's column_write_test.go.
 */
class ColumnWriteTest {

    private static Map<String, Object> row(Object... kv) {
        Map<String, Object> m = new LinkedHashMap<>();
        for (int i = 0; i < kv.length; i += 2) {
            m.put((String) kv[i], kv[i + 1]);
        }
        return m;
    }

    private static List<Map<String, Object>> items(Map<String, Object>... rows) {
        List<Map<String, Object>> out = new ArrayList<>();
        for (Map<String, Object> r : rows) {
            out.add(new LinkedHashMap<>(r));
        }
        return out;
    }

    private static final String[] MODES = {"row", "column"};

    @Test
    void bothModesWriteWholeColumn() throws Exception {
        for (String mode : MODES) {
            Frame f = Frame.create(mode, new HashMap<>(),
                items(row("id", "a"), row("id", "b"), row("id", "c")));
            OperatorOutput out = new OperatorOutput();
            out.setItemColumnDouble("score", new double[]{1.5, 2.5, 3.5});
            f.applyOutput(out, "op", false);

            double[] want = {1.5, 2.5, 3.5};
            for (int i = 0; i < want.length; i++) {
                assertEquals(want[i], f.item(i, "score"), mode + ": item " + i);
            }
            // Result projection includes the new field on every row.
            List<Map<String, Object>> result = f.toResultItems(List.of("id", "score"));
            for (int i = 0; i < result.size(); i++) {
                assertTrue(result.get(i).containsKey("score"),
                    mode + ": item " + i + " missing score in result");
            }
        }
    }

    @Test
    void lengthMismatchFails() {
        for (String mode : MODES) {
            Frame f = Frame.create(mode, new HashMap<>(), items(row("id", "a"), row("id", "b")));
            OperatorOutput out = new OperatorOutput();
            out.setItemColumnDouble("score", new double[]{1.0}); // wrong length
            Exception e = assertThrows(PineErrors.ExecutionError.class,
                () -> f.applyOutput(out, "op", false), mode);
            assertTrue(e.getMessage().contains("does not match item count 2"),
                mode + ": unexpected error: " + e.getMessage());
        }
    }

    @Test
    void nanValidationMessageParity() {
        for (String mode : MODES) {
            Frame f = Frame.create(mode, new HashMap<>(), items(row("id", "a"), row("id", "b")));
            OperatorOutput out = new OperatorOutput();
            out.setItemColumnDouble("score", new double[]{1.0, Double.NaN});
            Exception e = assertThrows(PineErrors.ExecutionError.class,
                () -> f.applyOutput(out, "op", false), mode);
            // Same first-error message shape as the per-element path.
            assertTrue(e.getMessage().contains(
                "item[1] write: field \"score\": NaN/Inf is not a valid JSON value"),
                mode + ": unexpected error: " + e.getMessage());
        }
    }

    @Test
    void columnWriteOverridesPerElement() throws Exception {
        for (String mode : MODES) {
            Frame f = Frame.create(mode, new HashMap<>(), items(row("id", "a"), row("id", "b")));
            OperatorOutput out = new OperatorOutput();
            out.setItem(0, "score", 99.0); // per-element first
            out.setItemColumnDouble("score", new double[]{1.0, 2.0});
            f.applyOutput(out, "op", false);
            // Column write applies after per-element → wins on collision.
            assertEquals(1.0, f.item(0, "score"), mode + ": column write must win");
        }
    }

    @Test
    void columnStoreAdoptsZeroCopy() throws Exception {
        // Column store must adopt the array (typed fast-path readable,
        // zero-copy) — pinned via the read view aliasing the written array.
        ColumnFrame f = new ColumnFrame(new HashMap<>(), items(row("id", "a"), row("id", "b")));
        double[] vals = {1.0, 2.0};
        OperatorOutput out = new OperatorOutput();
        out.setItemColumnDouble("score", vals);
        f.applyOutput(out, "op", false);

        double[] view = f.itemColumnDoubleView("score", 0, 2);
        assertNotNull(view, "adopted column must serve the typed read view");
        assertSame(vals, view, "expected zero-copy adoption (view aliases the written array)");
    }

    @Test
    void rowColumnResultParity() throws Exception {
        List<Map<String, Object>> src = items(
            row("id", "a", "old", 1.0), row("id", "b", "old", 2.0));
        Frame rowF = Frame.create("row", new HashMap<>(), items(row("id", "a", "old", 1.0), row("id", "b", "old", 2.0)));
        Frame colF = Frame.create("column", new HashMap<>(), src);
        for (Frame f : new Frame[]{rowF, colF}) {
            OperatorOutput out = new OperatorOutput();
            out.setItemColumnDouble("norm", new double[]{0.25, 0.75});
            f.applyOutput(out, "op", false);
        }
        List<String> fields = List.of("id", "old", "norm");
        assertEquals(rowF.toResultItems(fields).toString(), colF.toResultItems(fields).toString());
    }

    @Test
    void itemWriteMapFoldsColumnWrites() {
        OperatorOutput out = new OperatorOutput();
        out.setItem(0, "score", 99.0);
        out.setItem(1, "other", "x");
        out.setItemColumnDouble("score", new double[]{1.0, 2.0});
        Map<Integer, Map<String, Object>> m = out.itemWriteMap();
        assertEquals(1.0, m.get(0).get("score"), "column write overrides per-element in the view");
        assertEquals(2.0, m.get(1).get("score"));
        assertEquals("x", m.get(1).get("other"), "per-element writes on other fields preserved");
    }
}
